pkg/apis/edge/v1alpha1/edge-placement.go

```diff
@@ -102,6 +116,35 @@ type NonNamespacedObjectReferenceSet struct {
 	LabelSelectors []metav1.LabelSelector `json:"labelSelectors,omitempty"`
 }
 
+// UpsyncSet specifies a set of objects,
+// which may be namespaced or cluster-scoped,
+// from one particular API group.
+// An object is in this set if:
+// - its API group is the one listed;
+// - its resource (lowercase plural form of object type) is one of those listed;
+// - EITHER the resource is cluster-scoped OR the object's namespace matches `namespaces`; and
+// - the object's name matches `names`.
+type UpsyncSet struct {
+	// `apiGroup` is the API group of the referenced object, empty string for the core API group.
+	APIGroup string `json:"apiGroup,omitempty"`
+
+	// `resources` is a list of lowercase plural names for the sorts of objects to match.
+	// An entry of `"*"` means that all match.
+	// Empty list means nothing matches.
+	Resources []string `json:"resources"`
+
+	// `namespaces` is a list of acceptable namespaces.
+	// An entry of `"*"` means that all match.
+	// Empty list means nothing matches (you probably do not want this
+	// for namespaced resources).
+	Namespaces []string `json:"namespaces,omitempty"`
+
```

Comment:
Since we are still using TMC, and the TMC syncer (at least for now), namespaces names are generated by kcp and it may be quite difficult to specify selections based on namespaces. I think that until we create a proper EMC syncer that maintains namespace names from workload workspace to pCluster it will be pretty much impossible to use namespaces to filter.

Reply:
Any thoughts on how this up syncing selection mechanism would be implemented ? It looks like it will require a new controller within the syncer, perhaps in addition to the current upsyncer controller. Such controller will need to get the UpsyncSet spec. But such spec for now exists only in the Workload namespace. So I see couple of challenges here: 1) such controller would have to watch / read EdgePlacement objects in the Workload namespace (while currently only accesses the Mailbox workspace, and 2) such controller need to identify which EdgePlacement object applies to it.

Reply:
My first concern was defining something that makes sense when we get where we are trying to go, having an EMC syncer.  In that situation, the resolution of the upsync selection predicates to the objects to upsync would be done by the EMC syncer, and it would not be confounded by namespace mangling.

In the interim while we use the TMC syncer, can we turn off the namespace mangling?  I thought I heard that there were other users asking for this and @davidfestal willing to make it an option, but I might not be remembering correctly, but it could still be an option anyway.

In the interim while using the TMC syncer, one approach would be to supply our own controller that runs in the edge clusters to apply the labels needed to trigger upsync by TMC.  Our controller would need the relevant configuration information, which could relayed by the placement translator and the TMC syncer.  Is the namespace mangling predictable?  If so then the configuration for our controller could take it into account.

I still have questions relevant here about TMC upsync.  One is at https://kubernetes.slack.com/archives/C021U8WSAFK/p1678998135151379; getting an answer to that would help.

Another possible option would be to not bother implementing the mailbox/edge part of this while using the TMC syncer, instead making it the user's problem to make the objects-to-upsync at the edge have the labels that directly drive TMC upsync.  Evaluating the feasibility of this involves both getting the answer to my question above and understanding whether the C2P work needs upsync of only objects that are easy to get the needed labels on.

Reply:
> In the interim while using the TMC syncer, one approach would be to supply our own controller that runs in the edge clusters to apply the labels needed to trigger upsync by TMC. Our controller would need the relevant configuration information, which could relayed by the placement translator and the TMC syncer. Is the namespace mangling predictable? If so then the configuration for our controller could take it into account.

If I understand this correctly, that seems to best option, at least when using the TMC Syncer.
Yes, namespace mangling is predictable. Here is an example of code that retrieves it:

https://github.com/kcp-dev/kcp/blob/46f72457e994fabbda0175f57183ecc454b03a64/pkg/tunneler/podsubresourceproxy_handler.go#L204